teps/0004-tekton-integrations-annotations-statuses.md

```diff
@@ -0,0 +1,689 @@
+status data alongside IntegrationStatus data.
+
+Example:
+
+[CloudEventDelivery](https://pkg.go.dev/github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1?tab=doc#CloudEventDelivery):
+
+```yaml
+target: example.com
+status:
+  condition: Failed
+  sentAt: 123456789
+  error: "access denied"
+  retryCount: 1
+```
+
+IntegrationStatus:
+
+```yaml
+- name: cloudevent.integrations.tekton.dev
+  conditions:
+    - type: Succeeded
+      status: False
+      reason: "access denied"
+  lastTransitionTime: 123456789
+  annotations:
+    target: example.com
+    retryCount: 1
+```
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal and how do we mitigate. Think broadly.
+For example, consider both security and how this will impact the larger
+kubernetes ecosystem.
+
+How will security be reviewed and by whom?
+
+How will UX be reviewed and by whom?
+
+Consider including folks that also work outside the WGs or subproject.
+-->
+
+#### External Integration Quotas
+
+If relying on reconcilers / events that trigger on every update, integration
+providers will need to be careful to stay within quotas external providers may
+impose.
+
+While Tekton Pipelines will not impose specific throttling for integration
+providers, it will be a general best practice to use Status/Condition data (e.g.
+`ObservedGeneration`, `Conditions`, `LastTransitionTime`) as well as reconciler
+push back with `RequeueAfter` to stay within external quotas.
+
+### User Experience (optional)
+
+<!--
+Consideration about the user experience. Depending on the area of change,
+users may be task and pipeline editors, they may trigger task and pipeline
+runs or they may be responsible for monitoring the execution of runs,
+via CLI, dashboard or a monitoring system.
+
+Consider including folks that also work on CLI and dashboard.
+-->
+
+#### Storing data in Pipelines
+
+Similar to `TaskRunResults` and `PipelineResourceResults`, integration statuses
+are another form of external artifact generated as a result of the Pipeline.
+Because of this, it makes sense to store these statuses as a part of the
+Pipeline/TaskRun status, so that it is easy to find and see this data when
+something goes wrong.
+
+### Performance (optional)
+
+<!--
+Consideration about performance.
+What impact does this change have on the start-up time and execution time
+of task and pipeline runs? What impact does it have on the resource footprint
+of Tekton controllers as well as task and pipeline runs?
+
+Consider which use cases are impacted by this change and what are their
+performance requirements.
+-->
+
+N/A
+
+For pipelines, this should have no impact on runtime performance. Integration
+providers will need to be mindful of the information they store within the
+Pipeline/TaskRun objects to stay within k8s limits, but we are not exposing
+additional risk that did not already exist.
+
+## Design Details
+
+### Integration Status
```

Comment:
I have a little bit of trouble around this new field in status *and* how useful they are from the `PipelineRun`/`TaskRun` perspective. I would "tend" to think, that I would need to look at the "integration" service/dashboard/… to see what failed and dig from there.

In my head, I am comparing this to the "food supply tracability chain" :upside_down_face:. When you buy your cereals, you can trace back to where/when/how each element of your cereals got in (aka where was it planted, where was it transformed, by whom it was shipped, …). From the *top*, you can get to the *bottom* — but from the *bottom* you don't really have a way (nor the need) to get to the *top*. I see `PipelineRun` and `TaskRun` as "the bottom" in there (from the integration perspective).

Let's take GitHub status checks as examples, and let's assume a Github integration (à-la [`commit-status-tracker`](https://github.com/tektoncd/experimental/tree/master/commit-status-tracker)). I submit my PR, the checks appear as pending. If there is any failure (before the pod execution, during, …), it's the integration responsability to give the user the information back, pointing to what went wrong — and in there, which PipelineRun it is, it's state, how to dig into it, … The PipelineRun itself doesn't *need* to know what integration is watching it, and how it handles thing. 

*In addition, writing the status from the integration "part" means possible race, possible reconciling loop for all the watchers, etc…*

Reply:
So I actually agree that users will generally start from the integration dashboard and working their way back to the Run.

That said, I think providing information the other way is equally as important for investigating when things do not work as expected. For example, say the access token we are using for GitHub exceeded quota (or the token was revoked, GitHub is down, etc)- this could manifest as statuses never showing up in GitHub, so we lose that link from the integration back to the Run. We currently don't have a mechanism to let users know that something is wrong.
Today our answer is "go look at the controller logs for errors", but this isn't really a good solution if the user doesn't have access to the pipeline controller (and they likely won't if this is a multitenant environment). Having a place to say: "we tried to update the status for <x>, but we got this error" within a Run helps integrations give users transparency about what is happening as a result of their Task/PipelineRun, even though these things aren't necessarily executing as part of their Run, regardless of the state of the external integration.

I concede that in cases where everything worked this is less meaningful, but I think still provides benefit for consistency to help users verify that something actually happened.

To use your cereal analogy - this is like distributors keeping a log of outgoing shipments to grocery stores, so that if stores claim they never received a shipment, they can prove when the shipment left the warehouse, what was inside, tracking numbers, etc.